### ISSUE TITLE
Brute force knn tile size selection

### ISSUE THREAD

**Issue by tfeher** [rapidsai/cuvs#277]
**Describe the bug**

Heuristics incorrectly selects tiled execution of brute force-knn when the output tile could still fit the memory. This makes knn search slower than torch matmul + topk.

** Additional context **
 
https://github.com/rapidsai/cuvs/blob/72154b0b806c106300b52870f6113fdda3f87f0b/cpp/src/neighbors/detail/faiss_distance_utils.h#L50

**Steps/Code to reproduce bug**

Run brute force vector search using input with 1Mx128 input matrix and small number of queries. (The example below uses pylibraft, python wrappers, which currently has the same code as cuvs).

```python
import rmm
mr = rmm.mr.PoolMemoryResource(rmm.mr.CudaMemoryResource(), initial_pool_size=2**30)
rmm.mr.set_current_device_resource(mr)
import torch
import numpy as np
import pylibraft
from pylibraft.common import Handle
from pylibraft.neighbors.brute_force import knn
import cupy as cp
import time
device = torch.device("cuda")

class BenchmarkTimer:
    """Provides a context manager that runs a code block `reps` times
    and records results to the instance variable `timings`. Use like:
    .. code-block:: python
        timer = BenchmarkTimer(rep=5)
        for _ in timer.benchmark_runs():
            ... do something ...
        print(np.min(timer.timings))

        This class is part of the rapids/cuml benchmark suite
    """

    def __init__(self, reps=1, warmup=0):
        self.warmup = warmup
        self.reps = reps
        self.timings = []

    def benchmark_runs(self):
        for r in range(self.reps + self.warmup):
            t0 = time.time()
            yield r
            t1 = time.time()
            if r >= self.warmup:
                self.timings.append(t1 - t0)

rows = 1000000
cols = 128
n_queries = 8
k = 10
dataset = torch.randn(rows, cols, device=device)
queries = torch.randn(n_queries, cols, device=device)
dataset_cp = cp.asarray(dataset)
queries_cp = cp.asarray(queries)
timer = BenchmarkTimer(reps=100, warmup=5)
for rep in timer.benchmark_runs():
    distance = torch.matmul(queries, dataset.T)
    distances, indices = torch.topk(distance, k, dim=1, largest=True)

timings = np.asarray(timer.timings)
avg_time = timings.mean() * 1000
std_time = timings.std() * 1000
print("Average search time: {0:7.3f} +/- {1:7.3} ms".format(avg_time, std_time))

timer = BenchmarkTimer(reps=100, warmup=5)
handle = Handle()
for rep in timer.benchmark_runs():
    distances, indices = knn(dataset_cp, queries_cp, k=10, metric="sqeuclidean", handle=handle)
    handle.sync()

timings = np.asarray(timer.timings)
avg_time = timings.mean() * 1000
std_time = timings.std() * 1000
print("Average search time: {0:7.3f} +/- {1:7.3} ms".format(avg_time, std_time))

```

**Comment by tfeher** [comment]
cross referencing the FAISS issue where the problem was encountered https://github.com/facebookresearch/faiss/issues/3621

**Comment by jamxia155** [comment]
Running @tfeher's reproducer as is (with API updates for 26.04) gave me the following timings on an H100 node:
- torch:
`Average search time:   0.404 +/-   0.182 ms`
- cuvs: 
`Average search time:   1.384 +/-  0.0359 ms`

@tfeher pointed out that torch is essentially using the inner-product metric, which gives:
- cuvs with `metric="inner_product"`:
`Average search time:   0.527 +/- 0.00727 ms`

Investigations revealed 3 independent factors for cuvs performance deficit:

1. cuvs brute-force knn search currently [performs a shallow copy](https://github.com/rapidsai/cuvs/blame/a626f60f7b1a264b6b9f841036842841ff44a277/cpp/src/neighbors/detail/knn_brute_force.cuh#L490) of the cuvs resources handle. While this shallow copy increments the ref counters of existing resource shared pointers, any newly created resource is isolated from the original handle and gets destroyed at the end of the copy's lifetime. Pre-initializing the resources handle with the underlying resource needed improves timing:
`Average search time:   0.488 +/-  0.0056 ms`
2. cuvs-python allocates output `distances` and `neighbors` arrays if not supplied as input. torch is clearly not doing this allocation as nsys profile shows uninterrupted kernel activity throughout benchmarking iterations. Pre-allocating `distances` and `neighbors` for reuse gives:
`Average search time:   0.430 +/- 0.00574 ms`
3. Note that the torch timing above showed a large standard deviation (0.182 ms) relative to the mean (0.404 ms). It turns out that the torch timing is unexpectedly _fast_ for the first ~30 iterations (see 100 timings in ms below):
`0.121 0.121 0.122 0.124 0.13  0.123 0.123 0.122 0.121 0.12  0.119 0.121
 0.131 0.123 0.12  0.134 0.133 0.12  0.119 0.121 0.129 0.123 0.119 0.121
 0.122 0.12  0.12  0.119 0.129 0.327 0.514 0.514 0.517 0.516 0.511 0.519
 0.514 0.515 0.515 0.515 0.52  0.514 0.514 0.516 0.517 0.514 0.516 0.514
 0.517 0.515 0.515 0.516 0.516 0.514 0.515 0.516 0.516 0.513 0.515 0.513
 0.516 0.515 0.517 0.514 0.518 0.515 0.516 0.517 0.516 0.514 0.514 0.514
 0.518 0.516 0.515 0.516 0.516 0.515 0.513 0.515 0.517 0.519 0.515 0.515
 0.515 0.517 0.514 0.517 0.516 0.514 0.514 0.519 0.515 0.517 0.514 0.515
 0.517 0.515 0.515 0.514`
Note also that the extremely fast early iterations are NOT corroborated in nsys profiles. Increasing the warmup iterations from 5 to 50 gives:
`0.517 0.51  0.515 0.517 0.515 0.516 0.515 0.514 0.515 0.517 0.52  0.511
 0.514 0.514 0.516 0.517 0.513 0.514 0.517 0.516 0.515 0.516 0.511 0.517
 0.516 0.512 0.515 0.516 0.515 0.516 0.516 0.518 0.52  0.509 0.515 0.517
 0.513 0.515 0.514 0.518 0.517 0.513 0.515 0.516 0.516 0.517 0.514 0.517
 0.516 0.513 0.514 0.516 0.514 0.512 0.515 0.515 0.518 0.513 0.513 0.517
 0.516 0.511 0.517 0.516 0.516 0.513 0.515 0.515 0.515 0.516 0.516 0.517
 0.516 0.514 0.515 0.52  0.514 0.52  0.505 0.518 0.516 0.515 0.517 0.517
 0.514 0.522 0.508 0.517 0.517 0.514 0.516 0.516 0.514 0.516 0.515 0.517
 0.516 0.513 0.515 0.518`
`Average search time:   0.515 +/- 0.00241 ms`
It is not immediately clear why setting 5 warmup iterations causes inaccurate initial timings only for torch.

In summary, to regain performance advantage over torch, cuvs needs to:
- use the equivalent distance metric (inner product)
- reuse cuvs resources and memory allocations
- (not cuvs-related) use reliable timing instrumentation for benchmarking

**Comment by jamxia155** [comment]
Snippet for ensuring reuse of cublasLt handle:
```
# workaround: force-initialize cublaslt handle
from cuvs.distance import pairwise_distance
_ = pairwise_distance(cp.empty((1,1)), cp.empty((1,1)), metric="inner_product", resources=handle)
```
Snippet for pre-allocating output arrays before benchmarking loop:
```
distances = cp.empty((n_queries, k), dtype='float32')
indices = cp.empty((n_queries, k), dtype='int64')
```

**Comment by jamxia155** [comment]
Testing config for various batch sizes:
- cuvs: 26.04
- torch: 2.11.0+cu128
- Dataset size: 3M
- Dim: 128
- Datatype: float32
- Query batch size: variable

A100 PCIE 80GB:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jamxia/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jamxia/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">


batch_size | cuvs (ms) | torch (ms) | cuvs speedup
-- | -- | -- | --
1 | 1.112 | 1.147 | 1.03147482
10 | 1.915 | 2.501 | 1.306005222
100 | 7.155 | 11.822 | 1.652271139
1000 | 63.024 | 117.075 | 1.857625666
2000 | 127.357 | 237.944 | 1.868322903
3000 | 187.849 | 345.606 | 1.839807505
4000 | 253.15 | OOM |  
10000 | 667.181 | OOM |  



</body>

</html>


H100 PCIE 80GB:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jamxia/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jamxia/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">


batch_size | cuvs (ms) | torch (ms) | cuvs speedup
-- | -- | -- | --
1 | 1.185 | 1.199 | 1.011814
10 | 1.345 | 1.809 | 1.344981
100 | 4.409 | 7.842 | 1.778635
1000 | 42.343 | 82.787 | 1.955152
2000 | 84.884 | 165.086 | 1.944842
3000 | 121.865 | 247.882 | 2.03407
4000 | 167.814 | OOM |  
10000 | 418.098 | OOM |  



</body>

</html>
